kotlin/workflow-ui-android/src/main/java/com/squareup/workflow/ui/ModalContainer.kt

```diff
@@ -242,3 +255,26 @@ abstract class ModalContainer<ModalRenderingT : Any>
     )
   }
 }
+
+private class OnDestroy(private val block: () -> Unit) : LifecycleObserver {
+  @OnLifecycleEvent(ON_DESTROY)
+  fun onDestroy() = block()
+}
+
+private val Dialog.activity: AppCompatActivity?
+  get() = decorView?.context?.activity()
+
+private val Dialog.decorView: View?
+  get() = window?.decorView
+
+/**
+ * The [AppCompatActivity] for this context, or null if there isn't one.
+ *
+ * We keep all this very forgiving because we're just using it to keep some logging
+ * noise out of logcat. If someone manages to run this under the wrong type of
+ * activity, just return null and let the caller no-op.
+ */
+tailrec fun Context.activity(): AppCompatActivity? = when {
+  this is Activity -> this as? AppCompatActivity
+  else -> (this as? ContextWrapper)?.baseContext?.activity()
```

Comment:
This assumes there's only one context layer between the dialog and the activity. Is this guaranteed here? I've heard legends of a certain Square apps where there can be 10 layers of context and extracting the activity requires a while loop.

Reply:
and then I actually read the code and realized this is a recursive.. and tailrec so while loop. Oh well, all good.